src/main/java/org/apache/spark/shuffle/ucx/memory/MemoryPool.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+package org.apache.spark.shuffle.ucx.memory;
+
+import org.apache.spark.shuffle.UcxShuffleConf;
+import org.apache.spark.unsafe.Platform;
+import org.openucx.jucx.ucp.UcpContext;
+import org.openucx.jucx.ucp.UcpMemory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+/**
+ * Utility class to reuse and preallocate registered memory to avoid memory allocation
+ * and registration during shuffle phase.
+ */
+public class MemoryPool implements Closeable {
+  private static final Logger logger = LoggerFactory.getLogger(MemoryPool.class);
+
+  @Override
+  public void close() {
+    for (AllocatorStack stack: allocStackMap.values()) {
+      stack.close();
+      logger.info("Stack of size {}. " +
+          "Total requests: {}, total allocations: {}, preAllocations: {}",
+        stack.length, stack.totalRequests.get(), stack.totalAlloc.get(), stack.preAllocs.get());
+    }
+    allocStackMap.clear();
+  }
+
+  private class AllocatorStack implements Closeable {
+    private final AtomicInteger totalRequests = new AtomicInteger(0);
+    private final AtomicInteger totalAlloc = new AtomicInteger(0);
+    private final AtomicInteger preAllocs = new AtomicInteger(0);
+    private final ConcurrentLinkedDeque<RegisteredMemory> stack = new ConcurrentLinkedDeque<>();
+    private final int length;
+
+    private AllocatorStack(int length) {
+      this.length = length;
+    }
+
+    private RegisteredMemory get() {
+      RegisteredMemory result = stack.pollFirst();
+      if (result == null) {
+        ByteBuffer buffer = Platform.allocateDirectBuffer(length);
+        UcpMemory memory = null;
+        if (conf.preregisterMemory()) {
+          memory = context.registerMemory(buffer);
+        }
+        result = new RegisteredMemory(new AtomicInteger(1), memory, buffer);
+        totalAlloc.incrementAndGet();
```

Comment:
is this just for statistics?

Reply:
yes

Reply:
i'd suggest to remove it

Reply:
This makes easier to understand how many buffers were allocated so to use these statistics to preallocate them:
```
19/10/30 10:44:11,693 INFO MemoryPool: Stack of size 4096. Total requests: 205, total allocations: 25, preAllocations: 0
19/10/30 10:44:11,694 INFO MemoryPool: Stack of size 16777216. Total requests: 79, total allocations: 11, preAllocations: 0
19/10/30 10:44:11,694 INFO MemoryPool: Stack of size 8388608. Total requests: 111, total allocations: 15, preAllocations: 0
```
Then you can add `spark.shuffle.ucx.memory.preAllocateBuffers 4096:25:100,8388608:50` and you'll got:
```
19/10/30 10:44:11,693 INFO MemoryPool: Stack of size 4096. Total requests: 205, total allocations: 1, preAllocations: 1
19/10/30 10:44:11,694 INFO MemoryPool: Stack of size 16777216. Total requests: 79, total allocations: 1, preAllocations: 1
19/10/30 10:44:11,694 INFO MemoryPool: Stack of size 8388608. Total requests: 111, total allocations: 1, preAllocations: 1
```